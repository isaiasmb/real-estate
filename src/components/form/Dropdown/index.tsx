import React, { useState, useEffect } from 'react'
import Portal from 'components/Portal'
import * as S from './styles'

export type OptionType = {
  value: string
  label: string
  disabled?: boolean
}

type DropdownProps = {
  node: React.RefObject<HTMLDivElement>
  refId: string
  refIdDropdown: string
  options: OptionType[]
  selected: OptionType | null
  handleSelected: (value: OptionType | null) => void
  setOpenDropdown: (isOpenDropdown: boolean) => void
}

const Dropdown = ({
  node,
  refId,
  refIdDropdown,
  options,
  selected,
  handleSelected,
  setOpenDropdown
}: DropdownProps) => {
  const newNode = document.getElementById(refId)
  const [left, setLeft] = useState(0)
  const [top, setTop] = useState(0)
  const [minWidth, setMinWidth] = useState(0)

  useEffect(() => {
    if (newNode) {
      const rect = newNode.getBoundingClientRect()
      setLeft(rect.left)
      setTop(rect.top + rect.height + window.scrollY + 3)
      setMinWidth(rect.width)
    }
  }, [newNode])

  return (
    <Portal>
      <S.Wrapper
        top={top}
        left={left}
        minWidth={minWidth}
        ref={node}
        id={refIdDropdown}
      >
        <S.Options>
          {options.map((option) => (
            <S.Option
              key={option.value}
              id={option.value}
              selected={!!selected && selected.value === option.value}
              disabled={!!option.disabled}
              onClick={() => {
                if (option.disabled) {
                  return
                }

                handleSelected(option)
                setOpenDropdown(false)
              }}
            >
              {option.label}
            </S.Option>
          ))}
        </S.Options>
      </S.Wrapper>
    </Portal>
  )
}

export default Dropdown
