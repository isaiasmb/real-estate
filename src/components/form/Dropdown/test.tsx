import { fireEvent } from '@testing-library/react'
import React, { RefObject } from 'react'
import { renderWithTheme } from 'utils/tests/helpers'

import Dropdown, { OptionType } from '.'

const options: OptionType[] = [
  { value: 'brazil', label: 'Brazil', disabled: false },
  { value: 'italy', label: 'Italy', disabled: false },
  { value: 'ireland', label: 'Ireland', disabled: true }
]

describe('<Dropdown />', () => {
  let consoleSpy: jest.SpyInstance
  let refDropdown: RefObject<HTMLDivElement>
  let handleSelected: jest.Mock
  let setOpenDropdown: jest.Mock

  beforeEach(() => {
    if (consoleSpy) {
      consoleSpy.mockRestore()
    }

    refDropdown = React.createRef() as RefObject<HTMLDivElement>
    handleSelected = jest.fn()
    setOpenDropdown = jest.fn()
  })

  it('should render the dropdown with options', () => {
    consoleSpy = jest.spyOn(console, 'error').mockImplementation()

    const { container, rerender } = renderWithTheme(
      <div id="containerId"></div>,
      {
        container: document.body
      }
    )

    rerender(
      <Dropdown
        refId="containerId"
        refIdDropdown="dropdown"
        node={refDropdown}
        options={options}
        handleSelected={handleSelected}
        selected={null}
        setOpenDropdown={setOpenDropdown}
      />
    )

    expect(container.querySelectorAll('#portal-root ul')).toHaveLength(1)
    expect(container.querySelectorAll('#portal-root ul > li')).toHaveLength(3)
    container.querySelectorAll('#portal-root ul > li').forEach((item, key) => {
      const option = options[key]
      expect(item).toHaveProperty('id', option.value)
      expect(item).toHaveTextContent(option.label)

      if (option.value === 'ireland') {
        expect(item).toHaveStyle({ cursor: 'not-allowed' })
        expect(item).toHaveStyle({ backgroundColor: '#EAEAEA' })
      }
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should call handleSelected method when options is clicked', () => {
    consoleSpy = jest.spyOn(console, 'error').mockImplementation()

    const { container, rerender } = renderWithTheme(
      <div id="containerId"></div>,
      {
        container: document.body
      }
    )

    rerender(
      <Dropdown
        refId="containerId"
        refIdDropdown="dropdown"
        node={refDropdown}
        options={options}
        handleSelected={handleSelected}
        selected={null}
        setOpenDropdown={setOpenDropdown}
      />
    )

    fireEvent.click(container.querySelectorAll('#portal-root ul > li').item(0))

    expect(handleSelected).toBeCalled()
    expect(handleSelected).toBeCalledWith({
      value: 'brazil',
      label: 'Brazil',
      disabled: false
    })
  })

  it('should not call handleSelected method if the option is disabled', () => {
    consoleSpy = jest.spyOn(console, 'error').mockImplementation()

    const { container, rerender } = renderWithTheme(
      <div id="containerId"></div>,
      {
        container: document.body
      }
    )

    rerender(
      <Dropdown
        refId="containerId"
        refIdDropdown="dropdown"
        node={refDropdown}
        options={options}
        handleSelected={handleSelected}
        selected={null}
        setOpenDropdown={setOpenDropdown}
      />
    )

    fireEvent.click(container.querySelectorAll('#portal-root ul > li').item(2))

    expect(handleSelected).not.toBeCalled()
  })
})
