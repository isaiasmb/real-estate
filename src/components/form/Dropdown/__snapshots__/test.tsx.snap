// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should render the dropdown with options 1`] = `
.c0 {
  border-radius: 0.2rem;
  border: 0.1rem solid;
  border-color: #EAEAEA;
  position: absolute;
  z-index: 999999;
  top: 3px;
  left: 0px;
  min-width: 0px;
}

.c1 {
  overflow: auto;
  max-height: 230px;
  font-family: Raleway,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  color: #030517;
}

.c2 {
  padding: 8px 15px;
  cursor: pointer;
  background-color: #FAFAFA;
  color: #030517;
}

.c2:hover {
  background-color: #EAEAEA;
}

.c3 {
  padding: 8px 15px;
  cursor: pointer;
  background-color: #FAFAFA;
  color: #030517;
  cursor: not-allowed;
  background-color: #EAEAEA;
}

.c3:hover {
  background-color: #EAEAEA;
}

<div
  id="portal-root"
>
  <div>
    <div
      class="c0"
      id="dropdown"
    >
      <ul
        class="c1"
      >
        <li
          class="c2"
          id="brazil"
        >
          Brazil
        </li>
        <li
          class="c2"
          id="italy"
        >
          Italy
        </li>
        <li
          class="c3"
          disabled=""
          id="ireland"
        >
          Ireland
        </li>
      </ul>
    </div>
  </div>
</div>
`
